src/statsmodels/formula.jl

```diff
@@ -337,44 +336,46 @@ function StatsBase.model_response(mf::ModelFrame)
     end
 end
 
-modelmat_cols(v::DataVector) = asmatrix(convert(Vector{Float64}, v.data))
-modelmat_cols(v::Vector) = asmatrix(convert(Vector{Float64}, v))
 ## construct model matrix columns from model frame + name (checks for contrasts)
-function modelmat_cols(name::Symbol, mf::ModelFrame; non_redundant::Bool = false)
+function modelmat_cols{T<:AbstractFloatMatrix}(::Type{T}, name::Symbol, mf::ModelFrame; non_redundant::Bool = false)
     if haskey(mf.contrasts, name)
-        modelmat_cols(mf.df[name],
+        modelmat_cols(T, mf.df[name],
                       non_redundant ?
                       ContrastsMatrix{FullDummyCoding}(mf.contrasts[name]) :
                       mf.contrasts[name])
     else
-        modelmat_cols(mf.df[name])
+        modelmat_cols(T, mf.df[name])
     end
 end
 
+modelmat_cols{T<:AbstractFloatMatrix}(::Type{T}, v::DataVector) = convert(T, reshape(v.data, length(v), 1))
+modelmat_cols{T<:AbstractFloatMatrix}(::Type{T}, v::Vector) = convert(T, reshape(v, length(v), 1))
+
 """
-    modelmat_cols(v::PooledDataVector, contrast::ContrastsMatrix)
+    modelmat_cols(::Type{T}, v::PooledDataVector, contrast::ContrastsMatrix)
 
-Construct `ModelMatrix` columns based on specified contrasts, ensuring that
+Construct `ModelMatrix` columns of type `T` based on specified contrasts, ensuring that
 levels align properly.
 """
-function modelmat_cols(v::PooledDataVector, contrast::ContrastsMatrix)
+function modelmat_cols{T<:AbstractFloatMatrix}(::Type{T}, v::PooledDataVector, contrast::ContrastsMatrix)
     ## make sure the levels of the contrast matrix and the categorical data
     ## are the same by constructing a re-indexing vector. Indexing into
     ## reindex with v.refs will give the corresponding row number of the
     ## contrast matrix
     reindex = [findfirst(contrast.levels, l) for l in levels(v)]
-    return contrast.matrix[reindex[v.refs], :]
+    contrastmatrix = convert(T, contrast.matrix)
+    return contrastmatrix[reindex[v.refs], :]
 end
 
 """
     expandcols(trm::Vector)
```

Comment:
This signature should also be updated to mention the restriction on the element type.